### PR TITLE
Limit the addition of layers

### PR DESCRIPTION
**Pull request by gchoqueux**
## Description
Adjust the ability to add more color layers based on gpu's capabilities
and limit the addition of layers so that the abilities don't exceed

## Motivation and Context
Before the capabilities were clamp to 15 color layers
and if the limit is exceeded it will cause malfunction

**example if you exceed limit**
bug layer

### CODE REVIEW

File: src/Renderer/LayeredMaterial.js
Comment (autra) [inline src/Renderer/LayeredMaterial.js:81]: Unfortunately we cannot remove the `16-1` part yet, at least not on linux, not until https://bugzilla.mozilla.org/show_bug.cgi?id=777028 is fixed... :-/

Comment (autra) [reply src/Renderer/LayeredMaterial.js:81]: Oh forget this, you still have it on `arrayColorLength` I guess.

**Comment by peppsac** [comment]
I'm not a big fan of the changes proposed :(

IMHO using #309 as a starting point to reduce the number of used samplers would have been a better idea.
Then we could have improved it even further, by not tying canvas allocation and texture downloads. This way we could efficiently use GPU memory without pushing the complexity on the shaders.

File: src/Renderer/LayeredMaterial.js
Comment (gchoqueux) [reply src/Renderer/LayeredMaterial.js:81]: is with Mesa?

**Comment by gchoqueux** [comment]
@peppsac 
Texture atlas is good way to reduce  the number of used samplers
This PR isn't to reduce  the number of used samplers, it's to be able to use all the samplers.
the #309 doesn't resolve this problem, isn't it?
it would take a combination of both, ultimately

File: src/Renderer/LayeredMaterial.js
Comment (autra) [reply src/Renderer/LayeredMaterial.js:81]: Actually, I was right: on max, you can have `arrayColorLength * 2` samplers. So bug above applies.

**Comment by peppsac** [comment]
>  Texture atlas is good way to reduce the number of used samplers

Yes. Or to put it differently: to not waste samplers.

>  This PR isn't to reduce the number of used samplers, it's to be able to use all the samplers.

Sure. But IMHO it's way too convoluted, hence my suggestion to first simplify the code and then extend it.

File: src/Renderer/LayeredMaterial.js
Comment (gchoqueux) [reply src/Renderer/LayeredMaterial.js:81]: no, I don't think so
[because we test if we are in mesa driver](https://github.com/iTowns/itowns/pull/583/files#diff-dd48a35666ea82b037874482f9b61b60R23)

Comment (autra) [reply src/Renderer/LayeredMaterial.js:81]: Cool!

Comment (gchoqueux) [reply src/Renderer/LayeredMaterial.js:81]: but in fact it doesn't work
I am looking for a way to detect if mesa is present

**Comment by peppsac** [comment]
Another bothering issue with this change: you're adding a new layer of complexity that will only benefits to a portion of our users.
According to http://feedback.wildfiregames.com/report/opengl/feature/GL_MAX_TEXTURE_IMAGE_UNITS_ARB only 35% of users might have more than 16 active textures at a time - and keep in mind that this are statistics from a game so they may be skewed towards high-end graphics card.

And with all these changes, the total layer count will still be limited: even on an expensive GPU you won't be able draw all the Frecnh departments as geojson because that would require 101 layers...

So IMHO, we'd better focus on how to support an unlimited amout of color layers rather than adding code to just push the limit a bit further (and only for a fraction of users).

**Comment by nosy-b** [comment]
I think the first commit was needed: Avoiding the app to crash by not allowing to add too much layers (controling max textures per shader).

In real life we are talking about the number of textures per tile mesh which in the worst case we met  (projection between geometry tiles and textures tiles differs) is 3, meaning a mix of 5 different color layers per tileMesh which is for 99.9% of the users enough.

It is the job of the High level API to reorganize the layers if the user wants to add more than 5 color layer  by for example deactivating the rendering of the oldest one added. -> Not a job of itowns itself.

Clearly in the best world it would be better to handle an unlimited number of color layers together which would pass by texture atlas when we'll be able to make it as fast and light as our actual simple rendering architecture.

**Comment by peppsac** [comment]
@nosy-b sure the first commit implements a useful check. My comment is about the `SECOND_SAMPLER` changes.

**Comment by nosy-b** [comment]
@peppsac I don't feel that it looks much more complicated than before, we still got vec4 getColorAtIdUv(int id, vec2 uv)` and it looks quite transparent for most users.

Sure the architecture is not very elegant but it's mainly due to the openGL ES constraint: constant for loop and max number of _if_ statement. 

I join you on the fact that it might look like a 'lot' of code for a fraction of users so I'm not crazy about it but we can certainly imagine that the number of users with 32 texture samples available will grow in the future.

Another argument toward texture atlas ( or even mega texture :) ) is that we now make the assumption that we always have a single texture sample for the elevation but for some use cases this might not be enough...

**Comment by peppsac** [comment]
Then why not use something simpler that 2 manually unrolled for-loop?
The following code works for TEX_UNITS = 32 and is actually a simplification over the current iTowns code:
```glsl
vec4 colorAtIdUv(sampler2D dTextures[TEX_UNITS],vec3 offsetScale[TEX_UNITS],int id, vec2 uv){
    for (int i=0; i<TEX_UNITS; i++) {
        if (id == i) {
            return texture2D(dTextures[i],  pitUV(uv,offsetScale[i]));
        }
    }
    return vec4(0.0, 0.0, 0.0, 0.0);
}
```

**Comment by gchoqueux** [comment]
Some navigators doesn't support this type of for-loop (like IE)
In my memories, one or two years ago firefox or chrome didn't support it either.
We can't take the risk.

**Warning that appears with the addition of this method**
**_Firefox 52 / ubuntu 16.0_**
   * loop only executes for 1 iteration(s), consider removing [loop]

**_Google Chrome 62/ ubuntu 16.0_**
  * Can't unroll loops marked with loop attribute
  * Sampler array index must be a literal expression, forcing loop to unroll
	
**IE 12/ Windows 7**
 * error: shader can't be linked

**Comment by zarov** [comment]
Unfortunately, none of these versions are working with Mesa on Firefox with Ubuntu. The only working version is with the fallback from 32 to 16 [here](https://github.com/iTowns/itowns/pull/583/files#diff-dd48a35666ea82b037874482f9b61b60R36) **IF** we remove the condition on `/Sampler limit exceeded/`.

**Comment by gchoqueux** [comment]
@zarov I remove `/Sampler limit exceeded`/
But the roll for-loop, doens't work on Mesa/Ubuntu, because the index `i` isn't constant

File: src/Renderer/LayeredMaterial.js
Comment (gchoqueux) [reply src/Renderer/LayeredMaterial.js:81]: I fixe the firefox / mesa bug
unfortunately, the last test commit with roll loop doens't work in mesa

**Comment by gchoqueux** [comment]
branch updated

File: src/Core/System/Capabilities.js
Comment (zarov) [inline src/Core/System/Capabilities.js]: Why not put directly `return navigator.userAgent.toLowerCase().indexOf('firefox') > -1;`, or even `return navigator && navigator.userAgent && navigator.userAgent.toLowerCase().indexOf('firefox') > -1;` for the whole function ?

File: src/Renderer/LayeredMaterial.js
Comment (zarov) [inline src/Renderer/LayeredMaterial.js]: Can be removed

File: src/Renderer/Shader/TileFS.glsl
Comment (zarov) [inline src/Renderer/Shader/TileFS.glsl]: Space

**Review by zarov — COMMENTED** [review]
Looks good to me overall.

**Review by peppsac — DISMISSED** [review]
Still not convinced we want to add so much complexity, just for being able to add more color layers on few machines...
(and this PR will break panorama rendering because of missing `sumOccupancyColorLayers` function)

**Comment by peppsac** [comment]
There are code issues in the PR as well - but it's irrelevant for now: I don't like the solution proposed (for reasons already explained here: https://github.com/iTowns/itowns/pull/583#issuecomment-345948070)

**Comment by gchoqueux** [comment]
@peppsac 
* the addition of a sampler array adds little complexity. In addition there are getter and setter to make the use of textures completely transparent.

* 35% is a lot, it justifies to do it

* Indeed we all like to have the possibility to add an infinity of layer (within the limits of the capacity of the gpu). I will work on it at another time.

This PR doesn't add any bug or regression. So there's no real interest to block the PR

File: src/Core/Prefab/GlobeView.js
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: `wgs84TileLayer.sumOccupancyColorLayers = (colorLayers, layer) => {` is more correct.

File: src/Core/Prefab/GlobeView.js
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: Better avoid modifying arguments.
Change the signature of the function to `colorLayers` and let the caller passes the correct layer array.

File: src/Core/Prefab/GlobeView.js
Comment (peppsac) [reply src/Core/Prefab/GlobeView.js]: And an even better name would be: `countTextureSamplersForColorLayers`.
It's longer but it really describe what this function is doing.

Btw, this function (or its usage) isn't documented anywhere...

File: src/Core/Prefab/GlobeView.js
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: What is `getIncreColorlayer`?

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [inline src/Renderer/LayeredMaterial.js]: `getColorLayerCount`

Comment (peppsac) [reply src/Renderer/LayeredMaterial.js]: Or actually `getTextureSamplerCount`

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [inline src/Renderer/LayeredMaterial.js]: `getMaxTextureSamplerCount`?

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [inline src/Renderer/LayeredMaterial.js]: Remove dead code please.

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [inline src/Renderer/LayeredMaterial.js]: You could use multi-line syntax instead.

**Review by peppsac — COMMENTED** [review]
Removing my `Request Changes` tag as I'm the only one annoyed with those changes.

(I wrote a quick review nonetheless but only about cosmetic changes)

**Comment by gchoqueux** [comment]
fix #600

**Comment by zarov** [comment]
@peppsac did you want to remove your request for change ? It seems it's still here and locking this PR.

**Comment by gchoqueux** [comment]
[review](https://github.com/iTowns/itowns/pull/583#pullrequestreview-82833695) has been taken into account

**Comment by gchoqueux** [comment]
@iTowns/reviewers, the review has been taken into account

**Comment by gchoqueux** [comment]
we'll be able to merge this PR?

**Comment by gchoqueux** [comment]
@iTowns/reviewers this PR will be merged 23 January, if there's no request

**Comment by peppsac** [comment]
> @iTowns/reviewers this PR will be merged 23 January, if there's no request

Don't forget that every PR needs a review + approval.
I dismissed my review because I disagree with the ideas of this PR but that's different from an approval :)

So you still need to wait for one of the reviewers to perform a review.

**Comment by gchoqueux** [comment]
@peppsac thank you for finally answering
Don't worry, I haven't forgotten review + approval
I also asked if you are satisfied with the [cosmetic changes](https://github.com/iTowns/itowns/pull/583#pullrequestreview-82833695)

**Comment by autra** [comment]
> this PR will be merged 23 January, if there's no request

So you meant: "if there is a review", I guess ;-)

File: src/Core/View.js
Comment (zarov) [inline src/Core/View.js]: This is causing an error in `index.html` with the `menuGlobe` not being completed. Please adapt the example before merging.

**Review by zarov — COMMENTED** [review]
I'm not a fan of this solution, but if it means more layers for now until we implement atlases, I'm all for it. Next step is to make up-to-date #309 imo

File: src/Core/View.js
Comment (gchoqueux) [reply src/Core/View.js]: Ok PR Updated

**Comment by gchoqueux** [comment]
PR updated remove double array color texture

File: src/Renderer/LayeredMaterial.js
Comment (peppsac) [inline src/Renderer/LayeredMaterial.js:420]: `getTextureByIndex(l_COLOR, ...)` and `getTextureColorByIndex` are redundant from an API point of view.

`getTextureColorByIndex` shouldn't be exposed since its only a utility function.

And the same reasonning applies to all the new functions you're adding to LayeradMaterial: does it make sense to expose them?

File: src/Renderer/Shader/TileFS.glsl
Comment (peppsac) [inline src/Renderer/Shader/TileFS.glsl:15]: Good idea to rename all the variables to better names.

This could be improved though:
- texColor_01 -> colorTextures
- offsetScale_color_01 -> colorOffsetScale or colorTexturesOffsetScale
- paramLayers -> layerParameters or layerParams
- texElevation[1] -> elevationTexture
- offsetScale_elevation -> elevationTextureOffsetScale

And this change alone could be moved to a cleanup PR.

**Review by peppsac — CHANGES_REQUESTED** [review]
I just reviewed the renaming part.

And your last commit removes the initial idea of this PR: add a 2nd array of textures. Can you explain a bit more on this change?

**Comment by gchoqueux** [comment]
splited on 2 PRs : #782 and #784

**Pull request closed without merging**
